### ISSUE TITLE
[infra] Convert this mono_repo to use pub workspaces

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/native#1223]
From Dart 3.5 on, `pub` supports single resolution for related packages.

* https://flutter.dev/go/pub-workspace

This would be very useful for:

* `native_assets_cli`
* `native_assets_builder`
* `native_toolchain_c`

We can get rid of all the path dependencies / overrides.

Also, having a single resolution ensures that `package:jni` for example doesn't have any dependency conflicts with `package:native_assets_cli`.

It might be a slight over-specification to have _all_ packages in this repo as a single resolution, and it would mean bumping major versions needs to be done in PRs that touch all packages. But that might be a small price to pay.

We have to wait until Dart 3.5 reaches stable, the SDK constraint for the packages must be at least 3.5, and we don't want to push the packages to only work on beta/dev.

Thanks for the great work @sigurdm! ❤️ 

FYI @HosseinYousefi @liamappelbe @mosuem

**Comment by dcharkes** [comment]
We might also need the `resolution: local` feature, we have some example/test projects with a `pubspec.yaml` that are intentionally a different resolution:

* Flutter and non-Flutter test projects.
* Some example projects that explicitly target published older version of packages in this repo to ensure that version skew doesn't crash things.

@sigurdm What about projects that have `sdk: flutter` how does that relate to the requirement to have `sdk: 3.5.0`?

**Comment by sigurdm** [comment]
> What about projects that have sdk: flutter how does that relate to the requirement to have sdk: 3.5.0?

If you have `sdk:flutter` you need to have a flutter sdk with a dart that includes that specific version.
(You might get away with pointing `FLUTTER_ROOT` to a flutter repo if you just want to resolve)

**Comment by sigurdm** [comment]
`flutter` commands (flutter tools) still don't support workspaces. We'll have to wait for https://github.com/flutter/flutter/issues/150196

**Comment by Levi-Lesches** [comment]
@dcharkes Would you like a PR for this?

**Comment by dcharkes** [comment]
> @dcharkes Would you like a PR for this?

Yes please. It should work now that flutter is supporting pub workspaces as well.

Please note that we have a lot of test data which are projects as well, and some of those are explicitly pinned to _older_ versions of dependencies. For example pkgs/native_assets_builder/test/build_runner/version_skew_test.dart. I'm not entirely sure how that fits into pub workspaces. Please let me know if that causes trouble.

**Comment by sigurdm** [comment]
> I'm not entirely sure how that fits into pub workspaces.

As long as they are not added to the workspace explicitly, they should not cause trouble.

**Comment by Levi-Lesches** [comment]
Opened #1884, ready for review!
